### PR TITLE
add input value deprecation

### PR DESCRIPTION
**Pull request by asteinwedel**
Input value deprecation is a feature added a few years ago elsewhere that we would like to add support for. See [here](https://github.com/graphql/graphql-js/blob/main/src/utilities/getIntrospectionQuery.ts#L34) and [here](https://github.com/graphql/graphql-js/blob/7a6d055defd7a888f50c6133faae2b9010d45fad/src/type/directives.ts#L220) for graphql-js usage in introspection.

There are two main things I believe I still need to work on, one of which particularly I'd like some input on.
- ~~[x] add to deprecation tracking~~ removed
- [x] handling directive arg deprecation in various places~~ removed
- [x] fix a test

~~Regarding directives, when looking at tests in `DocumentAnalyzerSpec` and `IntrospectionSchemaMaterializerSpec`, I would like to test a field having a directive with a deprecated argument. However, the field only has `ast.Directive`s on it, which lack deprecation information (same for other types with deprecation, it's not in the `ast` objects).~~

~~It seems like I'd have to replace `Field.astDirectives: Vector[ast.Directive]` to instead be `Field.directives: Vector[schema.Directive]`. It is more align that way with `Field.arguments: List[schema.Argument]`, which has deprecation information. So it seems more in line with keeping `schema` type objects on itself.~~

~~I was wondering if that sounds like a good approach, or why `Field` has `ast.Directive` instead of `schema.Directive` in the first place?~~

~~(same goes for `InputField`, `Argument`, and various other places directives, being `ast.Directive` instead of `schema.Directive`)~~

answer @ `ast.Directive` vs `schema.Directive`, can usually find the corresponding `schema.Directive` on `Schema.directives`

### CODE REVIEW

**Comment by yanns** [comment]
Related to https://github.com/sangria-graphql/sangria/pull/992

**Comment by asteinwedel** [comment]
(also just updated something so working on tests)

> Related to #992

hahah should have looked first 🤦‍♀️

**Comment by asteinwedel** [comment]
@yanns an issue I have with https://github.com/sangria-graphql/sangria/pull/992 is it requires `isDeprecated` on `IntrospectionInputValue` and automatically adds it to the introspection query (can see in `graphql-js` that these are optional)

otherwise, I think we've mainly covered the same things though I've touched more places with checking deprecation

**Comment by yanns** [comment]
Yes I have the same issue with the introspection query.
I was thinking about adding a flag so that the user can decide whether to use it or not, but I never put time to finish this :D

**Comment by asteinwedel** [comment]
alright I added deprecation tracking and directive testing in various places

there is a failing test in `IntrospectionSchemaMaterializerSpec` in `testSchema` where `serverSchema` != `clientSchema`, but I'll have to check that out later

**Comment by asteinwedel** [comment]
@yanns I have tests and everything working now, but I'm a bit unfamiliar with how versioning works here

it looks like we're testing against sangria-core 2.13.12 and failing some checks

a few different things in this PR would be breaking changes, requiring a major version update, not sure what the process like that is here

**Comment by yanns** [comment]
First you can add the breaking changes in the build.sbt to that the build can run.
Depending on the number of breaking changes, I will decide for a major release or not.
Thanks for all your efforts!!!

**Comment by yanns** [comment]
There's one issue on Scala 3: https://github.com/sangria-graphql/sangria/actions/runs/6471747880/job/17571053945?pr=1063
strange

**Comment by asteinwedel** [comment]
> There's one issue on Scala 3: https://github.com/sangria-graphql/sangria/actions/runs/6471747880/job/17571053945?pr=1063 strange

hm i had a check to prevent this infinite recursion

I also noticed I'm missing something with nested input objects

so still looking at these things

q: what about splitting out deprecation tracking to a different PR? would it be okay to add support minus `DeprecationTracker` work? since that's a significantly larger change

**Comment by yanns** [comment]
Yes I wanted to ask why you have added deprecation tracking in this PR.
Isn't it something we can achieve with a middleware?
In general, yes to split this. 💯

**Comment by asteinwedel** [comment]
> Yes I wanted to ask why you have added deprecation tracking in this PR.
> Isn't it something we can achieve with a middleware?
> In general, yes to split this. 💯

no we aren't actually using the deprecation tracker so it's not as big of a deal for my purpose, I care more about supporting querying the introspection schema

I was adding it just to be consistent with other deprecation parts, but it's become quite big 😅 I'll split that out

File: modules/core/src/main/scala/sangria/introspection/model.scala [hunk 1]
```diff
@@ -137,4 +140,5 @@ case class IntrospectionDirective(
     description: Option[String],
     locations: Set[DirectiveLocation.Value],
     args: Seq[IntrospectionInputValue],
-    repeatable: Boolean)
+    repeatable: Boolean
+)
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/introspection/model.scala:144]: nitpick: this change is not necessary

File: modules/core/src/main/scala/sangria/schema/Schema.scala [hunk 2]
```diff
@@ -720,17 +721,22 @@ case class Argument[T](
     defaultValue: Option[(_, ToInput[_, _])],
     fromInput: FromInput[_],
     astDirectives: Vector[ast.Directive],
-    astNodes: Vector[ast.AstNode])
+    astNodes: Vector[ast.AstNode],
+    deprecationReason: Option[String])
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/schema/Schema.scala:725]: `Field` and `EnumValue` defines the fields in this order:
- `deprecationReason: Option[String]`
- `astDirectives: Vector[ast.Directive]`
- `astNodes: Vector[ast.AstNode]`

It'd be great to keep the same order for `InputField` and `Argument`.

File: modules/core/src/main/scala/sangria/schema/Schema.scala [hunk 3]
```diff
@@ -741,18 +747,21 @@ object Argument {
       name: String,
       argumentType: InputType[T],
       description: String,
-      defaultValue: Default)(implicit
+      defaultValue: Default
+  )(implicit
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/schema/Schema.scala:751]: Nitpick: this change is not necessary

File: modules/core/src/main/scala/sangria/schema/Schema.scala [hunk 4]
```diff
@@ -741,18 +747,21 @@ object Argument {
...
       toInput: ToInput[Default, _],
       fromInput: FromInput[T],
-      res: ArgumentType[T]): Argument[res.Res] =
+      res: ArgumentType[T]
+  ): Argument[res.Res] =
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/schema/Schema.scala:755]: Nitpick: this change is not necessary

File: modules/core/src/test/scala/sangria/introspection/IntrospectionSpec.scala [hunk 5]
```diff
@@ -536,7 +562,8 @@ class IntrospectionSpec extends AnyWordSpec with Matchers with FutureResultSuppo
                     "name" -> "includeDeprecated",
                     "description" -> null,
                     "type" -> Map("kind" -> "SCALAR", "name" -> "Boolean", "ofType" -> null),
-                    "defaultValue" -> "false")),
+                    "defaultValue" -> "false"
+                  )),
```
Comment (yanns) [inline modules/core/src/test/scala/sangria/introspection/IntrospectionSpec.scala:566]: Nitpick: this change is not necessary

File: modules/core/src/test/scala/sangria/introspection/IntrospectionSpec.scala [hunk 6]
```diff
@@ -582,7 +609,8 @@ class IntrospectionSpec extends AnyWordSpec with Matchers with FutureResultSuppo
...
```
Comment (yanns) [inline modules/core/src/test/scala/sangria/introspection/IntrospectionSpec.scala:613]: Nitpick: this change is not necessary

File: modules/core/src/test/scala/sangria/schema/SchemaComparatorSpec.scala [hunk 7]
```diff
@@ -203,13 +204,18 @@ class SchemaComparatorSpec extends AnyWordSpec with Matchers {
...
       """,
       nonBreakingChange[TypeAdded]("`Int` type was added"),
       breakingChange[InputFieldRemoved]("Input field `descr` was removed from `Filter` type"),
       nonBreakingChange[InputFieldAdded]("Input field `size` was added to `Filter` type"),
       nonBreakingChange[InputFieldDescriptionChanged]("`Filter.name` description is changed"),
-      nonBreakingChange[TypeDescriptionChanged]("`Filter` type description is changed")
+      nonBreakingChange[TypeDescriptionChanged]("`Filter` type description is changed"),
+      nonBreakingChange[InputFieldDeprecated]("`Filter.foo` was deprecated"),
+      nonBreakingChange[InputFieldAstDirectiveAdded](
+        "Directive `@deprecated` added on an input field `Filter.foo`")
```
Comment (yanns) [inline modules/core/src/test/scala/sangria/schema/SchemaComparatorSpec.scala:218]: 👍

File: modules/core/src/test/scala/sangria/schema/SchemaComparatorSpec.scala [hunk 8]
```diff
@@ -297,7 +305,11 @@ class SchemaComparatorSpec extends AnyWordSpec with Matchers {
       nonBreakingChange[ObjectTypeArgumentDefaultChanged](
         "`Filter.foo(a)` default value changed from none to `\"foo\"`"),
       nonBreakingChange[ObjectTypeArgumentDescriptionChanged](
-        "`Filter.foo(a)` description is changed")
+        "`Filter.foo(a)` description is changed"),
+      nonBreakingChange[ObjectTypeArgumentDeprecated](
+        "Argument `c1` on `Filter.foo` was deprecated"),
+      nonBreakingChange[FieldArgumentAstDirectiveAdded](
+        "Directive `@deprecated` added on a field argument `Filter.foo[c1]`")
```
Comment (yanns) [inline modules/core/src/test/scala/sangria/schema/SchemaComparatorSpec.scala:312]: 👍

File: modules/core/src/test/scala/sangria/schema/SchemaComparatorSpec.scala [hunk 9]
```diff
@@ -482,7 +494,10 @@ class SchemaComparatorSpec extends AnyWordSpec with Matchers {
       nonBreakingChange[DirectiveArgumentAstDirectiveRemoved](
         "Directive `@hello` removed from a directive argument `test.bar`"),
       nonBreakingChange[DirectiveArgumentAstDirectiveAdded](
-        "Directive `@world` added on a directive argument `test.bar`")
+        "Directive `@world` added on a directive argument `test.bar`"),
+      nonBreakingChange[DirectiveArgumentAstDirectiveAdded](
+        "Directive `@deprecated` added on a directive argument `test.foo`"),
+      nonBreakingChange[DirectiveArgumentDeprecated]("Directive argument `test.foo` was deprecated")
```
Comment (yanns) [inline modules/core/src/test/scala/sangria/schema/SchemaComparatorSpec.scala:500]: 👍

File: modules/derivation/src/main/scala/sangria/macros/derive/DeriveInputObjectSetting.scala [hunk 10]
```diff
@@ -9,6 +9,7 @@ case class InputObjectTypeDescription(description: String) extends DeriveInputOb
 
 case class DocumentInputField(fieldName: String, description: String)
     extends DeriveInputObjectSetting
+
```
Comment (yanns) [inline modules/derivation/src/main/scala/sangria/macros/derive/DeriveInputObjectSetting.scala:12]: Nitpick: this change is not necessary

**Review by yanns — COMMENTED** [review]
This is looking very good! Congrats for all new tests!

I left some minor comments to minimize the git diff.
The "real" comment is about the order of the fields.

**Comment by asteinwedel** [comment]
updated!

File: build.sbt [hunk 11]
```diff
@@ -211,7 +238,7 @@ lazy val sangriaCatsEffectExperimental = project
     name := "sangria-cats-effect-experimental",
     description := "Experimental support for Cats Effect",
     libraryDependencies ++= List(
-      "org.typelevel" %% "cats-effect" % "3.5.1",
+      "org.typelevel" %% "cats-effect" % "3.5.2",
```
Comment (yanns) [inline build.sbt:245]: Nitpick: this change is not related to the PR.

File: project/plugins.sbt [hunk 12]
```diff
@@ -2,6 +2,6 @@ addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "1.1.3")
 addSbtPlugin("org.scoverage" % "sbt-scoverage" % "2.0.9")
 addSbtPlugin("org.scoverage" % "sbt-coveralls" % "1.3.11")
 addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.4.6")
-addSbtPlugin("com.github.sbt" % "sbt-github-actions" % "0.16.0")
+addSbtPlugin("com.github.sbt" % "sbt-github-actions" % "0.17.0")
```
Comment (yanns) [inline project/plugins.sbt:5]: Nitpick: this change is not related to the PR.

**Review by yanns — COMMENTED** [review]
It's looking really good!
Question: why do you have picked 2 commits unrelated to this PR? (the dependency updates)

**Comment by asteinwedel** [comment]
ah sorry that was a rebase mistake

**Review by yanns — APPROVED** [review]
🥇 awesome, thank you for your contribution!

**Comment by asteinwedel** [comment]
@yanns could we get a release for this particularly? need to support introspection queries with this, could be a dev prerelease

**Comment by yanns** [comment]
You can try this pre-release: https://github.com/sangria-graphql/sangria/releases/tag/v4.1.0-RC1
It can take a few hours to synchronize with maven central.

**Comment by asteinwedel** [comment]
> You can try this pre-release: https://github.com/sangria-graphql/sangria/releases/tag/v4.1.0-RC1 It can take a few hours to synchronize with maven central.

thanks! works :)
